lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1462 @@
+                TIME_SECONDS (time));
+  else
+    g_snprintf (buffer,
+                sizeof (buffer),
+                "%d:%02d",
+                TIME_MINUTES (time),
+                TIME_SECONDS (time));
+
+  return (const gchar *) buffer;
+}
+
+/******************************** GST Support *********************************/
+static inline gboolean
+ekn_media_bin_handle_msg_error (EknMediaBin *bin, GstMessage *msg)
+{
+  GError *error = NULL;
+  gboolean handled;
+
+  gst_message_parse_error (msg, &error, NULL);
+
+  if (bin->play)
+    gst_element_set_state (bin->play, GST_STATE_NULL);
+
+  g_signal_emit (bin, ekn_media_bin_signals[ERROR], 0, error, &handled);
+
+  g_error_free (error);
+
+  return handled;
+}
+
+static inline void
+ekn_media_bin_update_duration (EknMediaBin *bin)
+{
+  gint64 duration;
+
+  gst_element_query_duration (bin->play, GST_FORMAT_TIME, &duration);
+
+  if (bin->duration == duration)
+    return;
+
+  bin->duration = duration;
+
+  duration = GST_TIME_AS_SECONDS (duration);
+  gtk_label_set_label (bin->duration_label, format_time (duration));
+  gtk_adjustment_set_upper (bin->playback_adjustment, duration);
+}
+
+static inline void
+ekn_media_update_position (EknMediaBin *bin)
+{
+  gint position = GST_TIME_AS_SECONDS (ekn_media_bin_get_position (bin));
+
+  if (bin->position == position)
+    return;
+
+  bin->position = position;
+  bin->ignore_adjustment_changes = TRUE;
+  gtk_adjustment_set_value (bin->playback_adjustment, position);
+  bin->ignore_adjustment_changes = FALSE;
+}
+
+static gboolean
+ekn_media_bin_tick_callback (GtkWidget     *widget,
+                             GdkFrameClock *frame_clock,
+                             gpointer       user_data)
+{
+  ekn_media_update_position ((EknMediaBin *)widget);
+
+  return G_SOURCE_CONTINUE;
+}
+
+static inline void
+ekn_media_bin_handle_msg_state_changed (EknMediaBin *bin, GstMessage *msg)
+{
+  GstState old_state, new_state;
+
+  gst_message_parse_state_changed (msg, &old_state, &new_state, NULL);
+
+  if (old_state == new_state ||
+      GST_MESSAGE_SRC (msg) != GST_OBJECT (bin->play))
+    return;
+
+  /* Update UI */
+  if (new_state == GST_STATE_PAUSED)
+    {
+      gtk_image_set_from_icon_name (bin->playback_image, EMB_ICON_NAME_PLAY, EMB_ICON_SIZE);
+      ekn_media_bin_update_duration (bin);
+    }
+  else if (new_state == GST_STATE_PLAYING)
+    {
+      gtk_image_set_from_icon_name (bin->playback_image, EMB_ICON_NAME_PAUSE, EMB_ICON_SIZE);
+      bin->tick_id = gtk_widget_add_tick_callback (GTK_WIDGET (bin), 
+                                                    ekn_media_bin_tick_callback,
+                                                    NULL, NULL);
+    }
+  else
+    {
+      gtk_image_set_from_icon_name (bin->playback_image, EMB_ICON_NAME_PLAY, EMB_ICON_SIZE);
+      bin->position = 0;
+      gtk_widget_remove_tick_callback (GTK_WIDGET (bin), bin->tick_id);
+      bin->tick_id = 0;
+    }
+}
+
+#if SHOW_METADATA
```

Comment:
Actually my plan is to make this information available somehow to the user with a keybinding or a right click option, so it wont be inside a #ifdef